README.md
# Crypto Ticker

## Overview

a TUI (Terminal User Interface) that displays 24 hour crypto tickers. 
Simply type a crypto symbol pair you are interest in and press either 
`F2` to add, or `F3` to remove its ticker from the table.

Ticker data is streamed via the <a href="https://github.com/binance/binance-spot-api-docs/blob/master/web-socket-streams.md" target="_blank">Binance socket streams</a> endpoint. Thus, 
supported symbol pairs directly correspond to the Binance WSS service. 
Please note that it may take several seconds for the stream to be added 
to the ticker table after pressing the `F2` button.

Press `F4` and `F5` to scroll the ticker table when you have added 
more rows than your terminal window can render.

![Crypto Ticker Demo](/doc/demo.gif)

## Application Details

This application is written in Haskell and makes use of the 
following concepts and libraries:

- **Clock**: For dealing with temporary data via the `clock` package.
- **Concurrency**: Multiple threads utilizing `TChan` and `BChan` primitives.
- **JSON**: Encoding and decoding JSON using the `aeson` package.
- **Lenses**: For accessing data in nested data structures.
- **TUI**: The terminal user interface is drawn with the `brick` library.
- **Web sockets**: Usage of the `Wuss` and `Network.WebSockets` modules.



CHANGELOG.md
# Revision history for proj-ticker

## 0.1.0.0 -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


test/tests.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE QuasiQuotes #-}

module Main where

import Test.Hspec
import Data.Aeson (encode, decode)
import Data.Maybe (fromJust)
import qualified Data.ByteString.Lazy as B
import Text.RawString.QQ
import WebTypes

-- -------------------------------------------------------------------
-- Test Functions

-- encoding a server request 
testEncodeSrq :: Bool
testEncodeSrq = encode sServerReq == sServerReqEnc

-- decoding a server request
testDecodeSrq :: Bool
testDecodeSrq = decode sServerReqEnc == Just sServerReq

-- encoding a server response 
testEncodeSrp :: Bool
testEncodeSrp = encode sServerResp == sServerRespEnc

-- decoding a server response 
testDecodeSrp :: Bool
testDecodeSrp = decode sServerRespEnc' == Just sServerResp

-- decoding a server error
testDecodeServerError :: Bool
testDecodeServerError = decode sServerErrorEnc == Just sServerError

-- decoding a ticker 
testDecodeTicker :: Bool
testDecodeTicker = decode sTickerEnc == Just sTicker

-- decoding a server message
testDecodeSrvMsg :: Bool
testDecodeSrvMsg =
  let r1 = MsgResponse (fromJust $ decode sServerRespEnc')   == MsgResponse sServerResp
      r2 = MsgError    (fromJust $ decode sServerErrorEnc) == MsgError sServerError
      r3 = MsgTicker   (fromJust $ decode sTickerEnc) == MsgTicker sTicker
  in r1 && r2 && r3

-- -------------------------------------------------------------------
-- Sample Data

-- encoded server request
sServerReqEnc :: B.ByteString
sServerReqEnc = [r|{"id":1,"method":"SUBSCRIBE","params":["btcusdt@miniTicker"]}|]

-- server request
sServerReq :: ServerRequest
sServerReq = ServerRequest
    { _srqRequestId = 1
    , _srqMethod    = "SUBSCRIBE"
    , _srqParams    = ["btcusdt@miniTicker"]
    }

-- server response
sServerResp :: ServerResponse
sServerResp = ServerResponse { _srpRequestId = 1 , _srpResult = [] }

-- encoded server response
sServerRespEnc, sServerRespEnc' :: B.ByteString
sServerRespEnc  = [r|{"result":[],"id":1}|]
sServerRespEnc' = [r|{"result":null,"id":1}|]

-- server error
sServerError :: ServerError
sServerError = ServerError 2 "invalid request."

-- encoded server error 
sServerErrorEnc :: B.ByteString
sServerErrorEnc = [r|{"code": 2, "msg": "invalid request."}|]

-- encoded ticker (encodes in alphabetical order by key)
sTickerEnc :: B.ByteString
sTickerEnc = [r|{
"E": 123456789,
"c": "0.0025",
"h": "0.0025",
"l": "0.0010",
"o": "0.0010",
"q": "18",
"s": "BTCUSDT",
"v": "10000"}|]

-- ticker sample
sTicker :: Ticker
sTicker = Ticker
  { _tckSymbolPair = "BTCUSDT"
  , _tckTs         = 123456789
  , _tckOpen       = 0.0010
  , _tckClose      = 0.0025
  , _tckHigh       = 0.0025
  , _tckLow        = 0.0010
  , _tckVolume     = 10000
  , _tckTrades     = 18
  }

-- -------------------------------------------------------------------
-- Main 

main :: IO ()
main = hspec $ do

  describe "JSON Encoding" $ do
    it "ServerRequest object encodes correctly" $ do
      testEncodeSrq

    it "ServerResponse object encodes correctly" $ do
      testEncodeSrq

  describe "JSON Decoding" $ do
    it "ServerRequest data decodes correctly" $ do
      testDecodeSrq

    it "ServerResponse data decodes correctly" $ do
      testDecodeSrp

    it "Server Error data decodes correctly" $ do
      testDecodeServerError

    it "Ticker data decodes correctly" $ do
      testDecodeTicker
    
    it "ServerMsg data decodes correctly" $ do
      testDecodeTicker
    


src/UI.hs
{-# LANGUAGE OverloadedStrings #-}
{-# OPTIONS_GHC -Wno-unrecognised-pragmas #-}
{-# HLINT ignore "Use ?~" #-}

module UI where

import Control.Concurrent.STM (newTChanIO)
import Control.Concurrent.STM.TChan (TChan)
import Brick
import qualified Brick.BChan as BC
import qualified Brick.Widgets.Border as B
import qualified Brick.Widgets.Center as C
import qualified Brick.Widgets.Edit as E
import qualified Brick.Focus as F
import qualified Brick.Types as BT

import Data.Maybe (fromJust)
import Control.Lens.TH
import Control.Lens ((^.), (.=), (%=))
import Lens.Micro.Mtl (use)

import Control.Monad.IO.Class (liftIO)
import Data.Text (Text)
import qualified Data.Text as T
import Data.Text.Zipper (clearZipper)
import qualified Network.WebSockets as WS
import qualified Graphics.Vty as V
import System.Clock (Clock(Monotonic), getTime)

import Format
import WebTypes
import UITypes
import Sockets
import Utils

-- -------------------------------------------------------------------
-- Event Handler

appEvent :: BrickEvent Name CustomEvent -> EventM Name AppState ()
appEvent e =
  case e of
    -- cache connection in app state
    AppEvent (CacheConnection c) -> conn .= Just c

    -- cache threadIds in app state
    AppEvent (CacheThreadId t) -> do
      threads %= \ts -> t : ts

    -- handle received table tickers
    AppEvent (TickerMessage t) -> do
      curTime <- liftIO $ getTime Monotonic

      -- update the removal buffer
      removeBuffer %= \rb -> if null rb then rb
        else updateRemoveBuffer curTime rb

      -- update ticker only if its not in the removal buffer
      rb <- use removeBuffer
      if isInRemoveBuffer (t^.tckSymbolPair) rb
        then return ()
        else
          -- sort tickers by symbol pair
          tickers %= \ts -> sortTickerAlpha $ rebuildTickers t ts

    -- wss error received
    AppEvent (ErrorMessage err) -> do
      processingReq .= False
      statusText .= (T.pack . show $ err)

    -- wss response received
    AppEvent (ResponseMessage res) -> do
      processingReq .= False
      statusText .= (T.pack . show $ res)

    -- tab keys (change focus)
    VtyEvent (V.EvKey (V.KChar '\t') []) ->
      focusRing %= F.focusNext

    VtyEvent (V.EvKey V.KBackTab []) ->
      focusRing %= F.focusPrev

    -- handle F2 key press
    VtyEvent (V.EvKey (V.KFun 2) []) -> do
      pr <- use processingReq
      e1 <- use edit1
      ts <- use tickers
      rb <- use removeBuffer

      let contents = head $ E.getEditContents e1
      let (passed, err) = validateAddRequest pr ts contents rb

      if not passed then statusText .= mkStatusText (fromJust err)
      else do
        -- write request
        processingReq .= True
        chan <- use reqChan
        liftIO $ Sockets.writeRequests [contents] "SUBSCRIBE" (fromJust chan)
        edit1 %= E.applyEdit clearZipper
        statusText .= mkStatusText SuccRequestSent `T.append` contents

    -- handle f3 key press
    VtyEvent (V.EvKey (V.KFun 3) []) -> do
      pr <- use processingReq
      ts <- use tickers
      e1 <- use edit1

      let contents = head $ E.getEditContents e1
      let (passed, err) = validateRemoveRequest pr ts contents
      
      if not passed then statusText .= mkStatusText (fromJust err)
      else do
        processingReq .= True
        let symbolLower = T.toLower contents
        chan <- use reqChan
        
        tickers %= removeTickerWithSymbol contents
        edit1 %= E.applyEdit clearZipper
        liftIO $ Sockets.writeRequests [symbolLower] "UNSUBSCRIBE" (fromJust chan)

        -- add symbol and current time to removal buffer
        curTime <- liftIO $ getTime Monotonic
        removeBuffer %= \rb -> (T.toUpper contents, curTime) : rb
        statusText .= contents `T.append` mkStatusText SuccRequestSent

    -- scroll viewport
    VtyEvent (V.EvKey (V.KFun 4) []) -> vScrollBy vp1Scroll 1

    VtyEvent (V.EvKey (V.KFun 5) []) -> vScrollBy vp1Scroll (-1)

    -- exit program
    VtyEvent (V.EvKey V.KEsc []) -> halt
    VtyEvent (V.EvKey (V.KChar 'q') []) -> halt

    -- handle editor widget
    ev -> do
      r <- use focusRing
      case F.focusGetCurrent r of
        Just Edit1 -> zoom edit1 $ E.handleEditorEvent ev
        Nothing    -> return ()
        _          -> return ()

-- -------------------------------------------------------------------
-- Viewport

vp1Scroll :: ViewportScroll Name
vp1Scroll = viewportScroll VP1

-- -------------------------------------------------------------------
-- UI

brickApp :: App AppState CustomEvent Name
brickApp = App { appDraw         = drawUI
               , appChooseCursor = appCursor
               , appHandleEvent  = appEvent
               , appStartEvent   = return ()
               , appAttrMap      = const theMap
               }

-- draw UI
drawUI :: AppState -> [Widget Name]
drawUI s = [ui s]

ui :: AppState -> Widget Name
ui st = if null (st ^. tickers)
         then C.center $ txt "connecting..."
         else
           vBox [ hBox [ vBox [ C.hCenterLayer (padTop (Pad 2)
                $ padBottom (Pad 2)
                $ drawEditor st
              <+> drawButtons st)
            ] ]
            -- vertical limit set to 80% of terminal height
            , vLimitPercent 80 $ viewport VP1 Vertical $ vBox [ padLeftRight 2 (drawTable st) ]
            , drawInfoLayer st
            ]

-- draw responsive table
drawTable :: AppState -> Widget Name
drawTable st = 
    vBox [ B.hBorder
         , mkHeaders
         , B.hBorder ] <=> drawRows ts

  where headers   = ["Symbols", " 24h%", "Open", "Close", "High", "Low", "Volume", "Traded"]
        cols      = map (withDefAttr (attrName "tableHeader") . padRight Max . txt) headers
        mkHeaders = hBox cols
        ts        = st ^. tickers

-- draw table rows
drawRows :: [Ticker] -> Widget Name
drawRows = foldr (\x acc ->
    let cs  = colText x
        cs' = map (padRight Max) cs
    in acc <=> vBox [ hBox cs', B.hBorder ]) emptyHBox
  where
    emptyHBox = hBox []
    colText t = [ txt $ t^.tckSymbolPair
                , colPct (pct t) . txt $ padPosPct (fmtPicoText 2 (pct t)) `T.append` "%"
                , txt $ fmtPicoText 4 (t^.tckOpen)
                , txt $ fmtPicoText 4 (t^.tckClose)
                , txt $ fmtPicoText 4 (t^.tckHigh)
                , txt $ fmtPicoText 4 (t^.tckLow)
                , txt $ fmtPicoText 2 (t^.tckVolume)
                , txt $ fmtPicoText 2 (t^.tckTrades)
                ]
      where diff t = 100 * (t^.tckClose - t^.tckOpen)
            pct t = diff t / (t^.tckOpen)

            -- style percentage values appropriately
            colPct v
              | abs v < 0.01 = withDefAttr (attrName "neutral")
              | v > 0        = withDefAttr (attrName "plusTicker")
              | otherwise    = withDefAttr (attrName "minusTicker")

            -- prefix space character to positive percentages
            padPosPct t
              | T.head t /= '-' = " " `T.append` t
              | otherwise       = t

-- edit box
drawEditor :: AppState -> Widget Name
drawEditor st = txt "Symbol Pair: " <+> hLimit 20 (vLimit 1 e1)
  where
    e1 = F.withFocusRing (st ^. focusRing)
                         (E.renderEditor (txt . T.unlines))
                         (st ^. edit1)

-- buttons
drawButtons :: AppState -> Widget Name
drawButtons st = padLeft (Pad 1) $ hBox $ padRight (Pad 1) <$> buttons
  where buttons = mkButton <$> buttonData
        buttonData = [ ("F2 ",    "Add",    attrName buttonAttr)
                     , ("F3 ",    "Remove", attrName buttonAttr)
                     , ("F4|F5 ", "Scroll", attrName buttonAttr)
                     ]
        buttonAttr = if st^.processingReq
                       then "buttonDisable"
                       else "buttonEnable"
        -- pass Name value to make button clickable
        mkButton (key, label, attr) =
          txt key <+>
          withDefAttr attr (padLeftRight 1 $ txt label)

-- draw status line
drawInfoLayer :: AppState -> Widget Name
drawInfoLayer st = Widget Fixed Fixed $ do
  c <- getContext
  let h = c ^. availHeightL
      msg = st ^. statusText
  render $ translateBy (Location (0, h-1))
         $ withDefAttr (attrName "info")
         $ C.hCenter
         $ txt msg

-- alternative function to draw status line
drawInfoLayer' :: AppState -> Widget Name
drawInfoLayer' st = withDefAttr (attrName "info")
                  $ hBox [ C.hCenter $ vLimit 1 $ txt msg ]
  where msg = st ^. statusText

-- -------------------------------------------------------------------
-- App Cursor 
-- https://hackage.haskell.org/package/brick-1.1/docs/Brick-Focus.html#v:focusRingCursor

appCursor :: AppState -> [BT.CursorLocation Name] -> Maybe (BT.CursorLocation Name)
appCursor = F.focusRingCursor (^.focusRing)

-- -------------------------------------------------------------------
-- Attribute Map

theMap :: AttrMap
theMap = theMap
  where
    theMap = attrMap V.defAttr
      [ (attrName "neutral",       fg V.brightBlue)
      , (attrName "plusTicker",    fg V.brightGreen)
      , (attrName "minusTicker",   fg V.brightRed)
      , (attrName "buttonEnable",  V.black `on` V.cyan)
      , (attrName "buttonDisable", V.black `on` V.color240 77 77 77)
      , (attrName "info",          V.white `on` V.color240 64 0 128)
      , (attrName "tableHeader",   V.defAttr `V.withStyle` V.bold)
      , (E.editAttr,               V.white `on` V.blue)
      , (E.editFocusedAttr,        V.white `on` V.blue)
      ]

-- -------------------------------------------------------------------
-- Main

initialState :: TChan ServerRequest -> AppState
initialState reqChan =
  AppState
  { _tickers       = []
  , _conn          = Nothing
  , _threads       = []
  , _reqChan       = Just reqChan
  , _focusRing     = F.focusRing [Edit1]
  , _edit1         = E.editor Edit1 (Just 1) ""
  , _statusText    = "F2/F3 to add/remove symbol pair"
  , _processingReq = False
  , _removeBuffer  = []
  }

main :: IO ()
main = do
  requestChannel <- newTChanIO
  eventChan <- BC.newBChan 10

  -- default state
  let s = initialState requestChannel

  -- spawn threads here
  connectBinance eventChan (fromJust $ s ^. reqChan)

  -- run ui
  let buildVty = V.mkVty V.defaultConfig
  initialVty <- buildVty
  finalState <- customMain initialVty buildVty
    (Just eventChan) brickApp s

  -- safely kill threads
  killThreads $ finalState ^. threads
  print "> killed threads"

  -- safely close connection
  WS.sendClose (fromJust $ finalState ^. conn) ("Bye!" :: Text)
  print "> killed connection"



src/Utils.hs
{-# LANGUAGE OverloadedStrings #-}

module Utils where 

import Control.Lens 
import Data.List (sortBy)
import Data.Text (Text)
import qualified Data.Text as T
import System.Clock (TimeSpec, toNanoSecs)
import WebTypes
import UITypes

-- -------------------------------------------------------------------
-- Functions called in event handlers

-- | Takes a symbol and checks if it is cached in the removal buffer.
isInRemoveBuffer :: Text -> [(Text, TimeSpec)] -> Bool
isInRemoveBuffer sym = foldr (\x acc ->
  if acc then acc else T.toUpper sym == fst x) False

-- | Deletes remove buffer items that have been cached for over 10 seconds
updateRemoveBuffer :: TimeSpec -> [(Text, TimeSpec)] -> [(Text, TimeSpec)]
updateRemoveBuffer curTime = filter (go curTime)
  where
    go ct x = let start   = snd x
                  nSecs   = toNanoSecs (ct - start)
                  seconds = fromIntegral nSecs / (10^9)
              in seconds < duration
    duration = 10.0  -- remove symbol from buffer after 10 seconds

-- | Sort tickers alphabetically based on the Ticker symbol pair
sortTickerAlpha :: [Ticker] -> [Ticker]
sortTickerAlpha = sortBy (\t1 t2 ->
  compare (t2^.tckSymbolPair) (t1^.tckSymbolPair))

removeTickerWithSymbol :: Text -> [Ticker] -> [Ticker]
removeTickerWithSymbol x = filter (\t -> (t^.tckSymbolPair) /= symUpper)
  where symUpper = T.toUpper x

-- | Validates editor text when adding a symbol pair
validateSymbol :: Text -> Bool
validateSymbol ts = isAlpha && noSpaces && goodLength
  where validChars = ['A'..'Z'] ++ ['a'..'z']
        isAlpha = foldr (\x acc ->
          if not acc then acc else x `elem` validChars) True (T.unpack ts)
        noSpaces = ' ' `notElem` T.unpack ts
        goodLength = T.length ts > 4

-- | Takes a symbol pair and returns True if corresponding Ticker exists
-- in the provided Ticker list.
searchTickerSymbol :: Text -> [Ticker] -> Bool
searchTickerSymbol t =
  foldr (\x acc -> if acc then acc else T.toUpper t == x ^. tckSymbolPair) False

-- | Takes a Ticker and returns True if a corresponding Ticker with the
-- same symbol pair exists in the provided Ticker list.
searchTicker :: Ticker -> [Ticker] -> Bool
searchTicker t = foldr (\x acc ->
    if acc then acc else x ^. tckSymbolPair == target) False
  where target = t ^. tckSymbolPair

-- | Updates a Ticker list by inserting the provided Ticker, or replacing
-- an older Ticker with the provided Ticker (with the same symbol pair).
rebuildTickers :: Ticker -> [Ticker] -> [Ticker]
rebuildTickers t ts
  | null ts = [t]
  | searchTicker t ts = map (\x -> if x ^. tckSymbolPair == target then t else x) ts
  | otherwise = t : ts
  where
    target = t ^. tckSymbolPair

-- | Status line text messages.
mkStatusText :: StatusLineMsg -> Text
mkStatusText ErrProcessingReq         = "A request is already being handled."
mkStatusText ErrTickerAlreadyAdded    = "The typed symbol is already added."
mkStatusText ErrValidationFailed      = "Validation failed. Enter a valid symbol pair."
mkStatusText ErrSymbolInRemovalBuffer = "Wait a few more seconds to add this symbol again."
mkStatusText ErrTickerListLength      = "At least one ticker must be present."
mkStatusText ErrNotInTickerList       = "The typed symbol pair is not in the ticker list."
mkStatusText SuccRequestSent          = "Request sent for "
mkStatusText SuccTickerRemoved        = " has been removed."

-- | Checks for some potential conditions that prevent removing
-- a ticker.
validateRemoveRequest :: Bool
                      -> [Ticker]
                      -> Text
                      -> (Bool, Maybe StatusLineMsg)
validateRemoveRequest processingReq ts sym
  | processingReq                   = (False, Just ErrProcessingReq)
  | length ts < 2                   = (False, Just ErrTickerListLength)
  | not (searchTickerSymbol sym ts) = (False, Just ErrNotInTickerList)
  | otherwise                       = (True, Nothing)

-- | Checks for some potential conditions that prevent adding
-- a ticker.
validateAddRequest :: Bool
                   -> [Ticker]
                   -> Text
                   -> [(Text, TimeSpec)]
                   -> (Bool, Maybe StatusLineMsg)
validateAddRequest processingReq ts sym rb
  | processingReq             = (False, Just ErrProcessingReq)
  | searchTickerSymbol sym ts = (False, Just ErrTickerAlreadyAdded)
  | not (validateSymbol sym)  = (False, Just ErrValidationFailed)
  | isInRemoveBuffer
    ( T.toUpper
    . T.strip $ sym) rb       = (False, Just ErrSymbolInRemovalBuffer)
  | otherwise                 = (True, Nothing)


src/Format.hs
{-# LANGUAGE OverloadedStrings #-}

module Format where 

import Control.Lens
import Data.Default
import Data.Text (Text, append, pack)
import qualified Formatting as F
import Data.Time.Clock.POSIX
import Data.Time.Clock

import WebTypes

(<^>) :: Text -> Text -> Text
(<^>) = append 

showt :: Show a => a -> Text
showt = pack . show

-- | Format any real number to `p` decimal points
-- Ex: fmtPicoText 4 0.00250012930
fmtPicoText :: Real a => Int -> a -> Text
fmtPicoText p = F.sformat (F.fixed p)

-- | Converts milliseconds to UTC time
-- Ex: fmtMsUTC 1661971736489
fmtMsUTC :: Integer -> UTCTime
fmtMsUTC t = posixSecondsToUTCTime $ fromInteger t / 1000

-- | Ouput a row of ticker information
fmtTickerRow :: Ticker -> Text
fmtTickerRow t = farRight
             <^> col (t ^. tckSymbolPair) "" <^> sep 
             <^> col (fmtPicoText 4 $ t ^. tckOpen)   "O" <^> sep 
             <^> col (fmtPicoText 4 $ t ^. tckClose)  "C" <^> sep 
             <^> col (fmtPicoText 4 $ t ^. tckHigh)   "H" <^> sep 
             <^> col (fmtPicoText 4 $ t ^. tckLow)    "L" <^> sep 
             <^> col (fmtPicoText 2 $ t ^. tckVolume) "V" <^> sep 
             <^> col (fmtPicoText 2 $ t ^. tckTrades) "T" <^> farLeft
  where col val p = shwPrefix p <^> val
        farRight = "| "
        farLeft  = " |"
        sep      = " | "
        shwPrefix p = if p /= "" then p <^> ": " else ""


src/Sockets.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE LambdaCase #-}

module Sockets where

import Control.Monad (forever, void, forM_)
import Control.Concurrent (forkIO, ThreadId, threadDelay, throwTo)
import Control.Concurrent.STM.TChan
import Control.Concurrent.STM (atomically, writeTChan)
import Control.Exception (AsyncException(ThreadKilled))
import Data.Aeson (encode, decode)
import Data.Text (Text)
import qualified Data.Text as T
import qualified Network.WebSockets as WS
import qualified Brick.BChan as BC
import qualified Wuss
import Formatting
import qualified System.Random as R
import WebTypes
import UITypes

-- -------------------------------------------------------------------
-- Web Sockets

-- | Initial symbol pairs to subscribe to
initialSymbolPairs :: [Text]
initialSymbolPairs = [ "btcusdt", "ethusdt", "adausdt"]

-- | Construct and Write ServerRequests to the request channel
writeRequests :: [Text] -> Text -> TChan ServerRequest -> IO ()
writeRequests symbols method chan = forM_ symbols $ \symbol -> do
  randId <- R.randomRIO (1,999)
  req    <- mkReqIO randId method (params symbol)
  atomically $ writeTChan chan req
  where
    params t = [t `T.append` "@miniTicker"]

-- | Connect to Binance WSS in a child thread
connectBinance :: BC.BChan CustomEvent -> TChan ServerRequest -> IO ()
connectBinance eventChan reqChan = do
  -- write initial symbol requests
  writeRequests initialSymbolPairs "SUBSCRIBE" reqChan

  -- spawn thread to handle wss connection
  void . forkIO $
    Wuss.runSecureClient url port path (wsApp eventChan reqChan)
    where
    url  = "stream.binance.com"
    port = 9443
    path = "/ws"

-- | App within web socket connection context
--
-- NOTE: This thread must stay alive until app is closed to
-- preserver the web socket connection.
wsApp :: BC.BChan CustomEvent -> TChan ServerRequest -> WS.ClientApp ()
wsApp eventChan reqChan conn = do

  -- spawn thread to receive responses
  t1 <- receiveWs eventChan conn
  
  -- cache connection and thread ids in app state
  BC.writeBChan eventChan (CacheConnection conn)
  BC.writeBChan eventChan (CacheThreadId t1)

  -- continues in the current thread
  sendWs conn reqChan

-- | Send data to websocket
sendWs :: WS.Connection -> TChan ServerRequest -> IO ()
sendWs conn reqChan =
  forever $ do
    -- read server request from channel
    msg <- atomically $ readTChan reqChan  
    WS.sendTextData conn (encode msg)

-- | Receive data from websocket
receiveWs :: BC.BChan CustomEvent -> WS.Connection -> IO ThreadId
receiveWs chan conn = do
  forkIO . forever $ do
    bs <- WS.receiveData conn
    let msg = decode bs
    forM_ msg $ \case
        MsgError    err -> BC.writeBChan chan (ErrorMessage err)
        MsgResponse res -> BC.writeBChan chan (ResponseMessage res)
        MsgTicker   t   -> BC.writeBChan chan (TickerMessage t)

-- | Kill threads
killThreads :: [ThreadId]-> IO ()
killThreads ts = do
  forM_ ts $ (flip throwTo) ThreadKilled

-- -------------------------------------------------------------------
-- Makers 

mkReq :: ServerRequest
mkReq = ServerRequest 1 "SUBSCRIBE" ["btcusdt@miniTicker"]

mkReqIO :: Int -> Text -> [Text] -> IO ServerRequest
mkReqIO reqId method params = return $ ServerRequest reqId method params

mkReqIO' :: IO ServerRequest
mkReqIO' = do
  rid <- R.randomRIO (1, 999)
  prompt "> Enter params: "
  params <- getLine
  return $ ServerRequest rid "SUBSCRIBE" [T.strip (T.pack params)]
  where
    prompt p = putStrLn p


src/UITypes.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE TemplateHaskell #-}

module UITypes where 

import Control.Concurrent (ThreadId)
import Control.Concurrent.STM.TChan
import Control.Lens.TH
import Data.Default
import Data.Text (Text)
import qualified Network.WebSockets as WS
import qualified Brick.Widgets.Edit as E
import qualified Brick.Focus as F
import Data.Map (Map)
import qualified System.Clock as C
import WebTypes

-- -------------------------------------------------------------------
-- Types

-- named resources 
data Name = Edit1 | VP1 | AddBtn | RemoveBtn  deriving (Ord, Show, Eq)

-- status line messages
data StatusLineMsg
  = ErrProcessingReq
  | ErrTickerAlreadyAdded
  | ErrValidationFailed
  | ErrSymbolInRemovalBuffer
  | ErrTickerListLength
  | ErrNotInTickerList
  | SuccRequestSent
  | SuccTickerRemoved

-- app state
data AppState = AppState
  { _tickers   :: [Ticker]
  , _conn      :: Maybe WS.Connection
  , _threads   :: [ThreadId]
  , _reqChan   :: Maybe (TChan ServerRequest)

    -- UI
  , _focusRing     :: F.FocusRing Name
  , _edit1         :: E.Editor Text Name
  , _statusText    :: Text
  , _processingReq :: Bool

    -- Remove Buffer 
    -- A removed symbol will be cached in this buffer for
    -- some time. Its used to prevent adding late Ticker data
    -- received from WSS after that ticker symbol has been
    -- removed.
    --
    -- The removed symbol cannot be re-added during the buffer.
    -- This will be communicated in the status line.
  , _removeBuffer  :: [(Text, C.TimeSpec)]
  }

data CustomEvent
  = CacheThreadId ThreadId          -- ^ for closing threads safely when app closes 
  | CacheConnection WS.Connection   -- ^ for closing connection safely when app closes 
  | ErrorMessage ServerError
  | ResponseMessage ServerResponse
  | TickerMessage Ticker
  | QuitApp

makeLenses 'AppState




src/WebTypes.hs
{-# LANGUAGE RecordWildCards #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE OverloadedStrings #-}

module WebTypes where 
import Control.Applicative ((<|>))
import Control.Lens.TH
import Data.Aeson
import Data.Default
import Data.Fixed (Pico)
import Data.Foldable (asum)
import Data.Maybe
import Data.Map (Map)
import Data.Text (Text, pack)
import GHC.Generics

type FixedFloat = Pico 

-- -------------------------------------------------------------------
-- Models

-- request 
data ServerRequest = ServerRequest
  { _srqRequestId :: Int            -- 1
  , _srqMethod    :: Text           -- SUBSCRIBE. UNSUBSCRIBE 
  , _srqParams    :: [Text]         -- "btcusdt@miniTicker"
  } deriving (Eq, Show)

-- responses 
data ServerResponse = ServerResponse
  { _srpRequestId :: Int            -- 1
  , _srpResult    :: [Text]         -- null (for subscribe, unsubscribe methods)
  } deriving (Eq, Show)

data ServerError = ServerError
  { _sveCode      :: Int            -- 2
  , _sveMessage   :: Text           -- "invalid..."
  } deriving (Eq, Show)

data Ticker = Ticker
  { _tckSymbolPair :: Text          -- "s", symbol,     "BTCUSDT"
  , _tckTs         :: Integer       -- "E", event time,  123456789
  , _tckOpen       :: FixedFloat    -- "o", open price,  0.0010
  , _tckClose      :: FixedFloat    -- "c", close price, 0.0025
  , _tckHigh       :: FixedFloat    -- "h", high price,  0.0025
  , _tckLow        :: FixedFloat    -- "l", low price,   0.0010
  , _tckVolume     :: FixedFloat    -- "v", total traded base asset volume, 10000
  , _tckTrades     :: FixedFloat    -- "q", total traded quote asset volume, 10
  } deriving (Eq, Show)

data ServerMsg
  = MsgResponse ServerResponse
  | MsgError    ServerError
  | MsgTicker   Ticker
  deriving (Eq, Show)


-- -------------------------------------------------------------------
-- JSON Encoding and Decoding

-- server request
instance ToJSON ServerRequest where
  toJSON (ServerRequest reqId method params) = object
    [ "id"     .= reqId
    , "method" .= method
    , "params" .= params
    ]

instance FromJSON ServerRequest where
  parseJSON = withObject "ServerRequest" $ \o -> do
    _srqRequestId <- o .: "id"
    _srqMethod    <- o .: "method"
    _srqParams    <- o .: "params"

    return ServerRequest{..}

-- server response
instance ToJSON ServerResponse where
  toJSON (ServerResponse resId result) = object
    [ "id"     .= resId
    , "result" .= result
    ]

instance FromJSON ServerResponse where
  parseJSON = withObject "ServerResonse" $ \o -> do
    _srpRequestId <- o .: "id"
    _srpResult    <- o .: "result" <|> pure []

    return ServerResponse{..}

-- server error
instance ToJSON ServerError where
  toJSON (ServerError code msg) = object
    [ "code" .= code
    , "msg"  .= msg
    ]

instance FromJSON ServerError where
  parseJSON = withObject "ServerError" $ \o -> do
    _sveCode    <- o .: "code"
    _sveMessage <- o .: "msg"

    return ServerError{..}

-- ticker
instance FromJSON Ticker where
  parseJSON = withObject "Ticker" $ \o -> do
    _tckSymbolPair <- o .: "s"
    _tckTs         <- o .: "E"
    _tckOpen       <- read <$> o .: "o"
    _tckClose      <- read <$> o .: "c"
    _tckHigh       <- read <$> o .: "h"
    _tckLow        <- read <$> o .: "l"
    _tckVolume     <- read <$> o .: "v"
    _tckTrades     <- read <$> o .: "q"

    return Ticker{..}

-- server message
instance FromJSON ServerMsg where
  parseJSON v = asum
    [ MsgResponse <$> parseJSON v
    , MsgError    <$> parseJSON v
    , MsgTicker   <$> parseJSON v
    ]

-- -------------------------------------------------------------------
-- Make Lenses 

makeLenses 'Ticker


-- -------------------------------------------------------------------
-- Defaults 

instance Default Ticker where
  def = Ticker { _tckSymbolPair = "BTCUSDT"
               , _tckTs         = 1661971736489
               , _tckOpen       = 0.0010
               , _tckClose      = 0.0025
               , _tckHigh       = 0.0045
               , _tckLow        = 0.0005
               , _tckVolume     = 10000
               , _tckTrades     = 10
               }
